transport/twirp/request_response_funcs.go

```diff
@@ -0,0 +1,41 @@
+package twirp
+
+import (
+	"context"
+	"github.com/twitchtv/twirp"
+	"net/http"
+)
+
+// ClientRequestFunc may modify the context. ClientRequestFuncs are executed
+// after creating the request but prior to sending the Twirp request to
+// the server.
+type ClientRequestFunc func(context.Context) (context.Context, error)
+
+// ServerRequestFunc may take information from the context. ServerRequestFuncs are
+// executed prior to invoking the endpoint.
+type ServerRequestFunc func(context.Context) context.Context
+
+// ServerResponseFunc may modify the context. ServerResponseFuncs are only executed in
+// servers, after invoking the endpoint but prior to writing a response.
+type ServerResponseFunc func(context.Context) (context.Context, error)
+
+// ClientResponseFunc may take information from the context. ClientResponseFuncs are only executed in
+// clients, after a request has been made, but prior to it being decoded.
+type ClientResponseFunc func(context.Context) (context.Context, error)
+
+// SetResponseHeader returns a ServerResponseFunc that sets the given header.
+func SetResponseHeader(key, val string) ServerResponseFunc {
```

Comment:
@basvanbeek what am I missing here in terms of your comment about http.Header?

Reply:
So SetResponseHeader is to report back headers to the client. What is missing is the ability to add headers to a client request and for a server to pick them up. Easiest is to think of 
`http.Header` to be similar to gRPC's `metadata.MD`. 

Client side you'd create a fresh empty http.Header which you'd pass to the server before funcs which can then add custom headers to this http.Header map potentially based on information found in client context. 

Server side you'd provide the received http headers from the client with a func that provides the server context and returns a potentially updated server context. This allows server before funcs to add stuff to context based on received metadata found in those client headers.